docs/tutorial_history_backend.rst

```diff
@@ -0,0 +1,319 @@
+So that we don't need any extra CouchDB's
+`Design Documents and Views <http://docs.couchdb.org/en/2.0.0/couchapp/ddocs.html>`_
+feature. Just with a bare ``_all_docs`` API, we can fetch history items back
+in order.
+
+Now that we have helper functions, we can update our ``append()`` method
+to do the real job - save history into DB.
+
+.. code-block:: python
+
+    def append(self, cmd):
+        self.inps.append(cmd['inp'])
+        self.rtns.append(cmd['rtn'])
+        self.outs.append(None)
+        self.tss.append(cmd.get('ts', (None, None)))
+        self._save_to_db(cmd)
+
+This method will be called by xonsh every time it run a new command from user.
+
+
+Retrieve History Items
+======================
+
+.. code-block:: python
+
+    def items(self):
+        yield from self._get_db_items(self.sessionid)
+
+    def all_items(self):
+        yield from self._get_db_items()
+
+These two methods are responsible for get history items for current xonsh
+session and all historical sessions respectively.
+
+And here is our helper methods to get docs from DB:
+
+.. code-block:: python
+
+    def _get_db_items(self, sessionid=None):
+        path = '/xonsh-history/_all_docs?include_docs=true'
+        if sessionid is not None:
+            path += '&start_key="{0}"&end_key="{0}-z"'.format(sessionid)
+        try:
+            r = self._request_db_data(path)
+        except Exception as e:
+            msg = 'error when query db: {}: {}'.format(e.__class__.__name__, e)
+            print(msg, file=sys.stderr)
+            return
+        data = json.loads(r.text)
+        for item in data['rows']:
+            cmd = item['doc'].copy()
+            cmd['ts'] = cmd['ts'][0]
+            yield cmd
+
+The `try-except` is here so that we're safe when something bad happened, like
+couchdb is not get started, etc.
+
+
+Try Out Our New History Backend
+===============================
+
+That's it. Your can find full code here:
+`<https://gist.github.com/mitnk/2d08dc60aab33d8b8b758c544b37d570>`_
+
+Let's start a new xonsh session:
+
+.. code-block:: none
+
+    $ history info
+    backend: couchdb
+    sessionid: 1486035364166-3bb78606-dd59-4679
+
+    $ ls
+    Applications   Desktop    Documents    Downloads
+
+    $ echo hi
+    hi
+
+Start a second xonsh session:
+
+.. code-block:: none
+
+    $ history info
+    backend: couchdb
+    sessionid: 1486035430658-6f81cd5d-b6d4-4f6a
+
+    $ echo new
+    new
+
+    $ history show all -nt
+    0:(2017-02-02 19:36) history info
+    1:(2017-02-02 19:36) ls
+    2:(2017-02-02 19:37) echo hi
+    3:(2017-02-02 19:37) history info
+    4:(2017-02-02 19:37) echo new
+
+    $ history -nt
+    0:(2017-02-02 19:37) history info
+    1:(2017-02-02 19:37) echo new
+    2:(2017-02-02 19:37) history show all -nt
+
+We don't miss any histories, so we're good I think.
+
+
+History Garbage Collection
+==========================
+
+In built-in history backends ``json``, ``sqlite``, GC will happen when
```

Comment:
"For the built-in history backends ``json`` and ``sqlite``, garbage collection is triggered when xonsh is started or when the user runs ``history gc``.  History items outside of the range defined by ``$XONSH_HISTORY_SIZE`` are deleted.